### ISSUE TITLE
Dart and analyzer produce different results for the cyclical typedef declaration

### ISSUE THREAD

**Issue by iarkh** [dart-lang/sdk#45117]
Dart SDK version: 2.13.0-59.0.dev (dev) (Sun Feb 21 18:06:58 2021 -0800) on "windows_x64"

Here is a source code example:
``` dart
class A<X extends A<X>> {}
typedef B<X extends A<A<X>>> = A<X>;

main() {}
```
Dart and analyzer produces different static errors here.

Sample output is:
```
$> dart --enable-experiment=nonfunction-type-aliases t.dart
t.dart:2:11: Error: Type argument 'A<X/*1*/>' doesn't conform to the bound 'A<X/*2*/>' of the type variable 'X' on 'A'.
 - 'A' is from 't.dart'.
 - 'X/*1*/' is from 't.dart'.
 - 'X/*2*/' is from 't.dart'.
Try changing type arguments so that they conform to the bounds.
typedef B<X extends A<A<X>>> = A<X>;
          ^
t.dart:1:9: Context: This is the type variable whose bound isn't conformed to.
class A<X extends A<X>> {}
        ^
t.dart:2:11: Error: Type argument 'X/*1*/' doesn't conform to the bound 'A<X/*2*/>' of the type variable 'X' on 'A'.
 - 'X/*1*/' is from 't.dart'.
 - 'A' is from 't.dart'.
 - 'X/*2*/' is from 't.dart'.
Try changing type arguments so that they conform to the bounds.
typedef B<X extends A<A<X>>> = A<X>;
          ^
t.dart:1:9: Context: This is the type variable whose bound isn't conformed to.
class A<X extends A<X>> {}
        ^
t.dart:2:32: Error: Type argument 'X/*1*/' doesn't conform to the bound 'A<X/*2*/>' of the type variable 'X' on 'A'.
 - 'X/*1*/' is from 't.dart'.
 - 'A' is from 't.dart'.
 - 'X/*2*/' is from 't.dart'.
Try changing type arguments so that they conform to the bounds.
typedef B<X extends A<A<X>>> = A<X>;
                               ^
t.dart:1:9: Context: This is the type variable whose bound isn't conformed to.
class A<X extends A<X>> {}
        ^

$> dartanalyzer --enable-experiment=nonfunction-type-aliases t.dart
Analyzing t.dart...
  error - 'A<X>' doesn't conform to the bound 'A<A<X>>' of the type parameter 'X'. - t.dart:2:23 - type_argument_not_matching_bounds
  error - 'X' doesn't conform to the bound 'A<X>' of the type parameter 'X'. - t.dart:2:25 - type_argument_not_matching_bounds
  error - 'X' doesn't conform to the bound 'A<X>' of the type parameter 'X'. - t.dart:2:34 - type_argument_not_matching_bounds
3 errors found.
```
As we can see form the output above, Dart claims about the line `2:11` and `2:32` whereas analyzer claims about the line `2:23`, `2:25` and `2:34`.

Seems like the both tools should behave in the similar manner.

It's affects co19 test [LanguageFeatures/Instantiate-to-bound/nonfunction_typedef/static/nonfunction_l1_t04.dart](https://github.com/dart-lang/co19/blob/master/LanguageFeatures/Instantiate-to-bound/nonfunction_typedef/static/nonfunction_l1_t04.dart)

**Comment by mit-mit** [comment]
cc @eernstg

**Comment by johnniwinther** [comment]
cc @stefantsov

**Comment by eernstg** [comment]
The considerations in [this comment](https://github.com/dart-lang/sdk/issues/45118#issuecomment-786509111) would apply here as well: We will not consider small differences in error location (or length) as bugs, as long as the _meaning_ of the error message can reasonably be claimed to be the same.

In this case, the errors are that (1) the bound `A<A<X>>` in `B` is not well-bounded (`X` cannot be used as a type argument to `A`, and `A<X>` cannot be used as a type argument to `A`), and (2) the body `A<X>` in `B` is not well-bounded (`X` cannot be used as a type argument to `A`). In both cases, we only have the assumption `X <: A<A<X>>`.

This is also the errors that are reported by the analyzer, and I think there's no reason to make this issue 'area-analyzer'.

For the CFE, however, I do think that it's a bit less readable when the error messages use the declarations rather than the applications to inform the developer about the bound violation:

```
... 'A<X/*1*/>' doesn't conform to the bound 'A<X/*2*/>' ...
 - 'X/*1*/' is from 't.dart'.
 - 'X/*2*/' is from 't.dart'.
```

I think we could have had something like the following, and also that this would be more readable:

```
... 'A<X>' doesn't conform to the bound 'A<A<X>>' ...
 - 'X' is from 't.dart' at line 2:13.
```

We would only have one `X` in this case, because we aren't showing the bound `A<X>` from line 1:19-or-so, we are showing the instantiation of the bound in the case where that `X` is bound to `A<X>`.

Another confusing detail is that the CFE reports an error on the bound, but it puts the location marker on the type parameter declaration (line 2:11 rather than line 2:23).

@stefantsov, do you agree that these changes could be an improvement? I'll keep the label 'area-front-end' and add 'improve-diagnostics' to reflect this perspective.

**Comment by stefantsov** [comment]
Yes, it's a long-standing issue of the CFE — not having the right offsets for errors involving types. In Kernel, the offsets aren't stored on `DartType`, so we point to the next best location, such as the type variable that has issues in its bound or the name of a method that has issues in its return type. I agree that changing the location for the error would be an improvement, and I think we should approach it on a case-by-case basis, as adding the offset to `DartType` probably would be too much of a change. I'll look into this particular case.
